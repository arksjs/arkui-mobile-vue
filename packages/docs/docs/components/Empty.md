# Empty 空状态

<CodeDemo name="Empty">

<<< @/../../demo/src/components/Show/Empty/index.vue

</CodeDemo>

## Import

```js
import { AkEmpty } from 'arkui-mobile-vue'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

### Import Type

组件导出的类型定义：

```ts
import type { EmptyType } from 'arkui-mobile-vue'
```

## Props

| 属性        | 类型      | 默认值    | 必填 | 说明                                                   |
| ----------- | --------- | --------- | ---- | ------------------------------------------------------ |
| description | string    |           | 否   | 描述文字                                               |
| type        | EmptyType | 'default' | 否   | 图片类型，可选 'default', 'error', 'network', 'search' |

## Slots

### 底部（#default）

```vue
<ak-empty description="网络错误" type="network">
  <ak-button>刷新试试</ak-button>
</ak-empty>
```

### 图片区域（#image）

```vue
<ak-empty description="描述文字">
  <template #image>
    <ak-image src="xxx.jpg" />
  </template>
</ak-empty>
```

注：优先级比内置图片高。
