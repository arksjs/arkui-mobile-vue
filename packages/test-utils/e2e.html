<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
