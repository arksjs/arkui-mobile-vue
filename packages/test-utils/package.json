{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
