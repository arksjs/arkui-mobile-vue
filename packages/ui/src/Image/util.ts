import { camelCase2KebabCase, getNumber, getEnumsValue, type CSSProperties } from '../helpers'
import type { Mode } from './types'

export const MODE_NAMES: Mode[] = [
  'scaleToFill',
  'aspectFit',
  'aspectFill',
  'widthFix',
  'top',
  'bottom',
  'left',
  'right',
  'top left',
  'top right',
  'bottom left',
  'bottom right'
]

export const getClasses = (autoHeight?: boolean) => [
  'ta-image',
  {
    'auto-height': !!autoHeight
  }
]

export const getImgClasses = (mode?: Mode, autoHeight?: boolean) => [
  'ta-image_img',
  'mode--' +
    camelCase2KebabCase(autoHeight ? 'widthFix' : getEnumsValue(MODE_NAMES, mode)).replace(
      /\s+/g,
      '-'
    )
]

export const getRatioStyles = (aspectRatio?: number | string) => {
  return {
    paddingTop: getNumber(aspectRatio) * 100 + '%'
  } as CSSProperties
}
