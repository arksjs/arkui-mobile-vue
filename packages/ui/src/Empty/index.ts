import { withInstall } from '../helpers/with-install'
import Empty from './Empty.vue'

export { Empty }
export const install = withInstall(Empty)
export default Empty
