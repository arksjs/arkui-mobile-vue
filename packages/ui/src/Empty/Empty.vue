<template>
  <div class="ta-empty">
    <slot name="image" v-if="$slots.image"></slot>
    <template v-else>
      <img class="ta-empty_image" :src="imageUrl" />
    </template>
    <p class="ta-empty_description" v-if="description">
      {{ description }}
    </p>
    <slot></slot>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue'
import type { PropType } from 'vue'
import { createEnumsValidator } from '../helpers/validator'
import type { EmptyType } from './types'
import { getImageUrl, TYPE_NAMES } from './util'

export default defineComponent({
  name: 'ta-empty',
  props: {
    // 描述文字
    description: {
      type: String,
      default: ''
    },
    // 类型
    type: {
      type: String as PropType<EmptyType>,
      validator: createEnumsValidator(TYPE_NAMES),
      default: TYPE_NAMES[0]
    }
  },
  setup(props) {
    const imageUrl = computed(() => getImageUrl(props.type))

    return {
      imageUrl
    }
  }
})
</script>
