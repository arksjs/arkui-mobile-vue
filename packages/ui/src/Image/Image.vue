<template>
  <div :class="classes" ref="root">
    <span
      v-if="!autoHeight && aspectRatio != null"
      class="ta-image_ratio"
      :style="ratioStyles"
    ></span>
    <i class="ta-image_loading" v-if="loading">
      <Icon :icon="loadingIcon" :size="iconSize" />
    </i>
    <i class="ta-image_error" v-if="error">
      <Icon :icon="errorIcon" :size="iconSize" />
    </i>
    <img v-if="currentSrc" :class="imgClasses" :src="currentSrc" @dragstart="onDrag" />
  </div>
</template>

<script lang="ts">
import {
  defineComponent,
  onMounted,
  ref,
  watch,
  onBeforeUnmount,
  computed,
  shallowRef,
  type PropType
} from 'vue'
import { Icon } from '../Icon'
import {
  addLazyQueue,
  loadNow,
  removeComponentFromLazy,
  withCheckInView
} from '../Image/load-image'
import {
  createEnumsValidator,
  emitErrorValidator,
  iconValidator,
  isString,
  type OnError,
  type PropsToEmits
} from '../helpers'
import type { ImageEmits, LoadedResource, Mode } from './types'
import ImageOutlined from '../Icon/icons/ImageOutlined'
import ImageBreakOutlined from '../Icon/icons/ImageBreakOutlined'
import type { IconData } from '../Icon/types'
import { getClasses, getImgClasses, getRatioStyles, MODE_NAMES } from './util'
import { useException } from '../hooks'

export default defineComponent({
  name: 'ta-image',
  components: { Icon },
  props: {
    // 图片资源地址
    src: {
      type: String,
      default: ''
    },
    mode: {
      type: String as PropType<Mode>,
      validator: createEnumsValidator(MODE_NAMES)
    },
    // 图片懒加载，在即将进入一定范围(preload=1.3)时才开始加载
    lazyLoad: {
      type: Boolean,
      default: false
    },
    // 自适应正方形
    aspectRatio: {
      type: [Number, String]
    },
    // 允许拖拽
    draggable: {
      type: Boolean,
      default: true
    },
    // 加载图标名称
    loadingIcon: {
      type: [String, Object] as PropType<IconData>,
      validator: iconValidator,
      default: ImageOutlined
    },
    // 加载失败图标名称
    errorIcon: {
      type: [String, Object] as PropType<IconData>,
      validator: iconValidator,
      default: ImageBreakOutlined
    },
    iconSize: {
      type: [Number, String]
    },
    // 自动计算高度
    autoHeight: {
      type: Boolean,
      default: false
    }
  },
  emits: {
    load: payload => payload && payload.width > 0 && payload.height > 0 && isString(payload.src),
    error: emitErrorValidator
  } as PropsToEmits<ImageEmits>,
  setup(props, { emit }) {
    const { createException } = useException()
    const loading = ref(true)
    const error = ref(false)
    const root = shallowRef<HTMLElement | null>(null)
    const currentSrc = ref<string | null>(null)
    const uid = Symbol()

    function load(src: string) {
      ;(props.lazyLoad ? addLazyQueue : loadNow)({
        src,
        uid,
        checkInView: withCheckInView(root.value as HTMLElement),
        onLoad,
        onError
      })
    }

    const onLoad = (res: LoadedResource) => {
      if (res.src === props.src) {
        // 防止多次变更图片导致的图片不正确
        loading.value = false
        error.value = false
        currentSrc.value = res.src
      }

      emit('load', {
        width: res.naturalWidth,
        height: res.naturalHeight,
        src: res.src
      })
    }

    const onError: OnError = e => {
      loading.value = false
      error.value = true

      emit('error', createException(e))
    }

    function onDrag(e: Event) {
      if (!props.draggable) {
        e.preventDefault()
      }
    }

    onMounted(() => props.src && load(props.src))

    onBeforeUnmount(() => removeComponentFromLazy(uid))

    watch(
      () => props.src,
      val => load(val)
    )

    const classes = computed(() => getClasses(props.autoHeight))
    const imgClasses = computed(() => getImgClasses(props.mode, props.autoHeight))
    const ratioStyles = computed(() => getRatioStyles(props.aspectRatio))

    return {
      currentSrc,
      classes,
      imgClasses,
      ratioStyles,
      loading,
      error,
      root,
      onDrag
    }
  }
})
</script>
