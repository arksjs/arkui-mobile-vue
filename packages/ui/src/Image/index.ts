import { withInstall } from '../helpers/with-install'
import Image from './Image.vue'

export { Image }
export const install = withInstall(Image)
export default Image
