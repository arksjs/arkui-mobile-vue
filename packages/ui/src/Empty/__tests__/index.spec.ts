import { describe, test, expect } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import { Empty } from '@/Empty'
import { Image } from '@/Image'

describe('Empty', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(Empty, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('description="description"', () => {
    const wrapper = shallowMount(Empty, {
      props: {
        description: 'description'
      }
    })

    expect(wrapper.find('.ta-empty_description').text()).toBe('description')
  })

  test('slot image', () => {
    const wrapper = shallowMount(Empty, {
      slots: {
        image: Image
      }
    })

    expect(wrapper.findComponent(Image).exists()).toBeTruthy()
  })
})
