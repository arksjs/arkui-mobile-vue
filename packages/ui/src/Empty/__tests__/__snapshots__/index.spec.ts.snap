// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Empty > snapshot 1`] = `
"<div class=\\"ta-empty\\"><img class=\\"ta-empty_image\\" src=\\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAUAAAAFACAMAAAD6TlWYAAABLFBMVEUAAAC5u70AAABWWWBVW2Dq6uy3ur1QV2AAAABWWmDv7+9WWmBXWmBWWmAAAAAAAABVWmAAAABVWWHr6+zr6+0AAAAAAADr6+1XW2BVWWBXWmBVWmDr6+sAAAAAAAAAAABXWmEAAAAAAADs7O1XWmAAAABWWWDs7O1VWmDt7e0AAABVWmDn5+4AAAC6u75WWWDr6+25ur63ub3s7OxVVWDt7e0AAAAAAAC4ury5vLxWXGBXWmHq6u8AAACvr7/s7O4AAAC5ur7X2Nq5ub2Nj5Sho6eFio+3t7/r6+y5ur319fVWWmDm5ud+gYXx8fHNztBfY2mNkJTY2dqqq6/f3+GQkpeusLKtr7JcYGbMzM+goqVvcnevsbTg4ePAwcTb292goqaHio13en9pbXFpGLi+AAAASHRSTlMAf3NAv2BgEKaPEN9f71UJ72wfv49kMu/Pv29vQBNAnH9HOd+fK6+fn29QMCAc79/Pv3BQMH98XURfUE8wJBCvkd+fj49/MCBUahNhAAAGiElEQVR42uzYzUrDQBSG4c+zkICQYRbNLiMJZBNopV1IXSgVFxK8//sxVilSG20yxLbhfS7hhXPmRwAAAAAAAAAAAAAAAAAAAMAZqEIuDFeZpcJw9QUHvHnrdKPWzLZmGtPMrNSFuusOePdfAXMzmwsxK9CE/srw6C59BZ5SZa00zJ03y4TeSvvCChzGFTPbSTMaDuCqbxF9Wjmht9zbThB6m5tZldXejMN4kMzMq5VnacqXwgCpWS0MZ9wCo1dgrhFl90lyH9QqvZ/gKVWO/FORJluFFMwu+NenUzArNKJNUjhXJLeSN+fMa2pcUTuN6PYz4EbylucTDDi2ItmqpZKr+iDh4xDJ1AreT28FAgAAAAAAAABwKutl85fFci10WDVHWQkHvS6aoyyehEPWzZEehEOumyMthY6AzHBcQGY4MiAzHBmQGY4MyAxHBmSGFR+Q10hEwOl6XukHAvbyon0E7GetPQRsndXannjAK+0hIAF/Q8BzQ0ACvrNzN61NRFEYxwcmZZgmgRIF09RFwBKVuBCqGxUEcXHJiZraRPElKW2//3doNsO5cKHkkHPp5DznWU22f8JvNWdE84Btmwc8oICb67mRLTePEvB6ZmbLRwk4n5nZ3AN6QMk8YNvmAT2gaB6wbfOAHlA0D9i2eUAPKFoLA1bTcsd9/dts8+PQFrIFnPYJYqt1poBDGnZ23PBfs9tvhzaiu5AnIFGFYOD6lm5y/QNLhICz9YLWWQLWNOYfx0cP7cufZvPvh7cV/c8S8AOdFs0ungW7u6JVloBVhOCTYHg/aZElYNFhBF8Ey1vQMkvAmurm8WWwvEv6nSVgyQgeBcu7okvdgCmCb4LlJQgmAR1BAYKKAcfQCCoELGkIjKBCwKqPjOA+ARnBj7gIxgEdQQGCmgFL6qIguMoSsGAEj4PlMYLKATvUg0VQJeCIEXwaLG+L4EMBHUERgmlAR1CEYBrQERQgqBtwRK9QEdQJOMVFcM+AjOAAFEGlgBNG8G2wvAjBNKAjuAuCv7IEHNBJ83gRLC9CMA3oCAoQ1A54TmeYCMYBHUERgmlAR1CAoHbAokvvIBHcPyAjCPGOESOoHLBHE0gE44BaCD4PlscIJgEdQRGCaUBHUICgfsAenSIiqBewwkRQISAjWAIiqBiwhkQwDugIChDUDMgIVngIKgYshoyg5YsbRlA7YE1jPARVAqZXX++D5cUIxgEdQQGCOQIWHUAEVQPWVMMhqBqwBERQNWBFBIdgHNARFCGYBnQEJQjqBkyvvl4Hy2MEdQNWfTgEtQKmV18gCCoHHMMhGAd0BEUIpgEdQRGCygH54AELQe2AIzQEtQOW1MVCUDtgwQh+Cpa3RTAJ6AjKEFQOmL7r+zlY3hbBLAGnYAgqBuSDByQE9QNOsBCMAzqCUgSzBJzSCRKCcUBHUIpgnoDndAaEoGZARhDjS1BbBLMEHEAheM/OHatEEARBGJ5lhWVwBMM9L9loEwVzc00aETHQ4IQT3/8dbE7m3NSmK6qqR/iDDxa2JzlgR5DiObwTgtuAQjCAICJgs5UHwW1AIfhvBA/pAfvVFw2C24BCMIAgJGCzHQ2CkICTXXO8BOUIQgI6giMLgpiA1fYUz+E5gtuAQjCA4AskoCM4kSB4hAQsMw2CX5iA1RYSBD8wAR94EDz8BRSCEQRfIQHLTHL//2bPmIDVKhuCuQFHlvt/RxAScDJjQ3AoqbugQ3AoQjCIYA8oBIMIQgJOV2wIJgbsV19cCGYHXKxSPIzsCGICjjazIPjeA6YjSPEwsiMICegINioE0wPekCGYHnC0SyoEh5I9MgRPAYVgEMEeUAjGEMQEvOVCMDdg/9eX4v7fEcQEXKkQBAT0gweK+39HsAcUglEEe0AhGEQQE9D/9eVB8DegEAwiiAnoBw88CJ4DCsEQgpiAjiDFm6COICZgs5UGQUhAR5DjYWRHMD1gv/piQbAHFIJBBM8BhWAMQUzAiQfBx9InBEMILgWy1SrF5/ARFXA02xMg+PltrWC2GMfu7wpobTaC7Z6KpmmapmmapmnaT3t1jMIgEEVR9DOVEJhgoZ0MCDZpLK1cgfvfTyYGUqQQTGOKc5Zwi/cAAAAAAAAA4B/dtgO3qLr00QXf5u3ALCAAwPXWnFKeosqlLMFJY9r1EUtT5eCcLvX3e5/aiNIMQ1OCc9p3wE7AH/VpN0bkpjKCp02vE1mjWkp5BAAAAAAAAADAlZ5X9riGUcwcQgAAAABJRU5ErkJggg==\\">
  <!--v-if-->
</div>"
`;
